doc/architecture.md

```diff
@@ -0,0 +1,174 @@
+# `go-substrate-gen` Architecture
+## Goal
+The overall goal of `go-substrate-gen` is to generate code that exposes methods to access all storage items and generate `go-substrate-rpc-client` calls for all extrinsics.
+As all substrate storage items and extrinsics are per-pallet, we generate them per-pallet, and contain them in a directory structure that looks like the below, if `runtime` is the name of the folder the user put all the generated code within
+```
+runtime/
+    types/
+        types.go
+    pallet1
+        calls.go
+        storage.go
+    pallet2
+        calls.go
+        storage.go
+    ...
+```
+The user can then call methods within their pallets by importing them from those go files
+```golang
+import (
+    ...
+    "github.com/path/to/package/runtime/pallet1"
+    "github.com/path/to/package/runtime/types"
+    ...
+)
+
+// Create calls to use with the rpc client
+// Using example pallet with extrinsic 'TransferAll'
+call := pallet1.MakeTransferAllCall(..).AsCall()
+
+// Get storage items
+// Using storage item named 'Balance'
+datatype := pallet1.GetBalanceLatest(..).AsCall()
+```
+which requires
+- a type for every storage item could be retrieved
+- an appropriate type for each extrinsic call that could be submitted
+
+`go-substrate-gen` uses [Jennifer](https://github.com/dave/jennifer) to perform the actual code generation. 
+
+## Overview of Code Structure
+The overall approach used is very simple:
+
+1. Parse the metadata returned from the `state.getMetadata` RPC endpoint by any substrate chain.
+1. Create a type generator which caches generated types
```

Comment:
1,1,1?

Reply:
I guess that is a little ugly. Markdown numbers it properly anyway, if you look at the rendered version, but I guess I'll swap that over